Index: sys/boot/forth/beastie.4th
===================================================================
--- sys/boot/forth/beastie.4th	(revision 256917)
+++ sys/boot/forth/beastie.4th	(working copy)
@@ -151,21 +150,20 @@
 
 	3 + \ beastie adjustment (see `fbsdbw-logo' comments above)
 
-	2dup at-xy ."  ```                        `" 1+
-	2dup at-xy ." s` `.....---.......--.```   -/" 1+
-	2dup at-xy ." +o   .--`         /y:`      +." 1+
-	2dup at-xy ."  yo`:.            :o      `+-" 1+
-	2dup at-xy ."   y/               -/`   -o/" 1+
-	2dup at-xy ."  .-                  ::/sy+:." 1+
-	2dup at-xy ."  /                     `--  /" 1+
-	2dup at-xy ." `:                          :`" 1+
-	2dup at-xy ." `:                          :`" 1+
-	2dup at-xy ."  /                          /" 1+
-	2dup at-xy ."  .-                        -." 1+
-	2dup at-xy ."   --                      -." 1+
-	2dup at-xy ."    `:`                  `:`" 1+
-	2dup at-xy ."      .--             `--." 1+
-	     at-xy ."         .---.....----."
+	2dup at-xy ."                              " 1+
+	2dup at-xy ."                             " 1+
+	2dup at-xy ."                             " 1+
+	2dup at-xy ."                 ______         " 1+
+	2dup at-xy ."                /      \        " 1+
+	2dup at-xy ."          _____/    f   \       " 1+
+	2dup at-xy ."         /     \        /       " 1+
+	2dup at-xy ."        /   p   \______/  Sense " 1+
+	2dup at-xy ."        \       /      \        " 1+
+	2dup at-xy ."         \_____/        \       " 1+
+	2dup at-xy ."               \        /       " 1+
+	2dup at-xy ."                \______/        " 1+
+	2dup at-xy ."                             " 1+
+	     at-xy ."                             "
 
  	\ Put the cursor back at the bottom
  	0 25 at-xy
