diff --git a/sys/net/if_bridge.c b/sys/net/if_bridge.c
index 572ee06..6e060f0 100644
--- a/sys/net/if_bridge.c
+++ b/sys/net/if_bridge.c
@@ -815,6 +815,8 @@ bridge_ioctl(struct ifnet *ifp, u_long cmd, caddr_t data)
 		}
 		BRIDGE_LOCK(sc);
 		LIST_FOREACH(bif, &sc->sc_iflist, bif_next) {
+			if (bif->bif_ifp->if_type == IFT_GIF)
+				continue;
 			if (bif->bif_ifp->if_mtu != ifr->ifr_mtu) {
 				log(LOG_NOTICE, "%s: invalid MTU: %lu(%s)"
 				    " != %d\n", sc->sc_ifp->if_xname,
@@ -1103,6 +1105,7 @@ bridge_ioctl_add(struct bridge_softc *sc, void *arg)
 	}
 #endif
 	/* Allow the first Ethernet member to define the MTU */
+	if (ifs->if_type != IFT_GIF) {
 	if (LIST_EMPTY(&sc->sc_iflist))
 		sc->sc_ifp->if_mtu = ifs->if_mtu;
 	else if (sc->sc_ifp->if_mtu != ifs->if_mtu) {
@@ -1110,6 +1113,7 @@ bridge_ioctl_add(struct bridge_softc *sc, void *arg)
 		    ifs->if_mtu, ifs->if_xname, sc->sc_ifp->if_mtu);
 		return (EINVAL);
 	}
+	}
 
 	bif = malloc(sizeof(*bif), M_DEVBUF, M_NOWAIT|M_ZERO);
 	if (bif == NULL)
