diff --git a/sys/boot/i386/boot0/boot0.S b/sys/boot/i386/boot0/boot0.S
index 798ee97..ab17088 100644
--- a/sys/boot/i386/boot0/boot0.S
+++ b/sys/boot/i386/boot0/boot0.S
@@ -647,8 +647,8 @@ os_dos:
 #endif
 os_win:		.ascii "Wi";   .byte 'n'|0x80
 os_linux:	.ascii "Linu"; .byte 'x'|0x80
-os_freebsd:	.ascii "Free"
-os_bsd:		.ascii "BS";   .byte 'D'|0x80
+os_freebsd:	.ascii "pfSe"
+os_bsd:		.ascii "ns";   .byte 'e'|0x80
 #ifndef SAVE_MORE_MEMORY
 os_ext:		.ascii "EX";   .byte 'T'|0x80
 #endif
