Index: sys/boot/forth/loader.conf
===================================================================
--- sys/boot/forth/loader.conf	(revision 256917)
+++ sys/boot/forth/loader.conf	(working copy)
@@ -529,3 +529,5 @@
 #module_before="cmd"		# executes "cmd" before loading the module
 #module_after="cmd"		# executes "cmd" after loading the module
 #module_error="cmd"		# executes "cmd" if load fails
+
+loader_color="NO"
