Index: sys/boot/forth/menu.4th
===================================================================
--- sys/boot/forth/menu.4th	(revision 256917)
+++ sys/boot/forth/menu.4th	(working copy)
@@ -445,7 +445,7 @@
 
 	\ Print the frame caption at (x,y)
 	str_loader_menu_title getenv dup -1 = if
-		drop s" Welcome to FreeBSD"
+		drop s" Welcome to pfSense"
 	then
 	TRUE ( use default alignment )
 	str_loader_menu_title_align getenv dup -1 <> if
