Index: sys/boot/forth/menu-commands.4th
===================================================================
--- sys/boot/forth/menu-commands.4th	(revision 256917)
+++ sys/boot/forth/menu-commands.4th	(working copy)
@@ -247,7 +247,7 @@
 
 	cr
 	." To get back to the menu, type `menu' and press ENTER" cr
-	." or type `boot' and press ENTER to start FreeBSD." cr
+	." or type `boot' and press ENTER to start pfSense." cr
 	cr
 
 	FALSE \ exit the menu
